[김선호] 1단계 - OAuth2.0 로그인 리뷰 요청드립니다 🚀

haero77 opened the pull request:
진영님 안녕하세요! 만나서 반갑습니다 ☘️

이번 미션의 경우 시큐리티 코드를 뜯어보면서 제 것으로 만들 수 있는 부분만 가져와서 약간은 제 입맛대로 구현해봤는데요,
시큐리티 철학에 크게 어긋나거나 한 부분이 있을지 같이 봐주시면 감사하겠습니다.

고민했던 내용과 궁금한 점은 코멘트로 남겨두겠습니다. 리뷰 잘 부탁드려요..! 🙇‍♂️

haero77 committed:
fix: FilterChainProxy가 필터 체인에 등록되어 SecurityFilterChain이 중복으로 실행되는 버그 수정

- FilterChainProxy는 빈으로 등록 안 하고, DelegatingFilterProxy만 빈으로 등록하여 서블릿 필터 체인에는 FilterChainProxy이 등록되지 않게 함으로써 해결

haero77 committed:
(step1-1) OAuth2 인증에 사용되는 정보를 환경변수로 관리

haero77 committed:
(step1-1) 깃헙 로그인 요청 시, 깃헙 로그인 페이지로 리다이렉트한다

haero77 committed:
(step1-1) docs: 요구사항 리스트업

haero77 committed:
(step1-1) 깃헙 로그인 이후 발급 받은 code를 추출하여 인증 매니저에게 인증을 위임한다

haero77 committed:
(step1-1) OAuth 환경변수 관리 컴포넌트를 security 패키지로 이동

haero77 committed:
(step1-1) 깃헙에서 발급받은 code를 액세스 토큰으로 교환한다

haero77 committed:
(step1-1) oauth2 관련 클래스를 security.oauth2 패키지로 응집

haero77 committed:
(step1-1) 깃헙에서 발급받은 액세스 토큰으로 유저 정보를 요청한다

haero77 committed:
(step1-1) ClientRegistration를 내부 클래스에서 일반 클래스로 격상

haero77 committed:
(step1-1) 유저 정보 조회 후 신규 회원 여부에 따라 회원가입 및 인증 처리

haero77 committed:
(step1-2) docs: 요구사항 리스트업

haero77 committed:
(step1-2) 리소스 서버로부터 사용자 리소스를 조회하는 로직을 DefaultOAuth2UserService로 추출

haero77 committed:
test: 테스트 간 데이터 오염 방지

- InmemoryMemberRepository 의 정적 변수를 인스턴스 변수로 변경
- 각 테스트 클래스간 다른 스프링 컨텍스트 사용하도록 테스트 구조 변경

haero77 committed:
(step1-2) 구글 클라이언트 환경변수를 객체로 관리

haero77 committed:
(step1-2) 구글 리다이렉트 필터 구현

haero77 committed:
(step1-2) OAuth2 제공자에 맞는 OAuth2AuthenticationToken을 생성하도록 리팩토링

haero77 committed:
(step1-2) OAuth2 Provider의 token URI와 user info URI를 환경변수로 관리

haero77 committed:
(step1-1) GitHub Authorization scope read:user로 변경

haero77 committed:
(step1-2) 액세스 토큰, OAuth2 유저 정보를 조회하는 책임을 ProviderClient로 위임

haero77 committed:
(step1-2) OAuth2 관련 클래스를 성격에 따라 oauth2.authentication, oauth2.provider 패키지로 응집

haero77 committed:
(step1-2) 구글 OAuth2 인증 프로세스 구현

haero77 committed:
docs: update Requirements, add Questions, Review Request, Step 1 Architecture image

haero77 commented on src/main/java/nextstep/app/security/CustomOAuth2UserService.java:
```diff
@@ -0,0 +1,45 @@
+        return new CustomOAuth2User(newMember.getEmail());
+    }
+
+    private String generateUsername(OAuth2UserRequest userRequest, OAuth2User oAuth2User) {
+        return userRequest.registration().provider().getProviderName() + "_" + oAuth2User.getName();
+    }
```
### _OAuth2 인증을 통해 서비스에 회원가입하게 될 경우, username을 어떻게 만드는 것이 좋을지에 대한 고민_

각 OAuth2 제공자마다, 식별할 수 있는 값이 주어지기 때문에 이것을 이용하여 username을 만들면 되겠다는 생각을 했습니다. 다만 OAuth2 제공자가 다르지만, 식별자가 같을 수도 있으므로(예를 들면 GitHub의 식별자와 Google의 식별자가 같은 경우) 식별자를 그대로 username으로 쓰기보다는 OAuth2 제공자를 Prefix로 하여 username을 생성했습니다. 

일반적으로 이런 방법을 채택하는지 궁금하네요. 또, 리뷰어님께서는 OAuth 인증 구현시 username을 주로 어떤 방식을 택하여 구현하시는지 궁금합니다 ㅎㅎ

haero77 replied:
### _OAuth2 인증 후 임의로 생성된 username으로 로그인을 허용할 것인가에 대한 고민_  

현재 서비스 로직에서는 OAuth2 인증 후 회원가입 과정을 거치게 됩니다.
그런데 이 때 회원가입을 위해 Member를 생성해야하고, 그래서 username을 일단 생성할 수밖에 없습니다.

그럼 이 때 임의로 생성되는 username에 대해, 이 username을 이용하여 로그인 시도를 했을 때 로그인을 허용할지에 대해 고민이 되더라고요. 이건 유저가 만든 것이 아니고 서버가 임의적으로 생성한 것이니까 유저가 인지하지 못하는 상황이고, 따라서 로그인을 어쨌든 막아두는게 맞지 않을까하는 생각이 들었습니다.

물론 유저가 인지하지도 못한 값이고, 비밀번호도 현재는 랜덤으로 생성해서 인증 오류가 발생하겠지만, 
(유저가 인지하지 못하는) 서버가 생성한 username으로 로그인할 때 웹 서버는 이것을 어떻게 처리하는게 좋을지 리뷰어님의 의견이 궁급합니다..!

haero77 commented on src/main/java/nextstep/security/oauth2/provider/GoogleClient.java:
```diff
@@ -0,0 +1,92 @@
+package nextstep.security.oauth2.provider;
+
+import nextstep.security.authentication.AuthenticationException;
+import nextstep.security.oauth2.authentication.ClientRegistration;
+import nextstep.security.oauth2.authentication.OAuth2AccessToken;
```
_**패키지 양방향 참조에 대한 고민**_

oauth2 클래스가 많아지다보니 구조화를 위해 먼저 oauth2 관련된 클래스를 oauth2 패키지로 두고, 
각 OAuth2 제공사마다 달라져야하는 로직을 oauth2.provider로, 나머지는 oauth2.authentication으로 위치시켜봤습니다.

다만 슬슬 구현이 끝날 때쯤 보니 provider와 authentication이 패키지 레벨에서 봤을 떄 양방향 참조가 존재하게 되더라고요. 기존에 패키지 레벨의 양방향 참조가 순환참조로 발전될 수 있다고 인지하고 있던 참이라, 패키지 뎁스를 더 두어서 아예 패키지 레벨에서의 양방향 참조를 없애버리는게 나을지 고민됩니다..!

jinyoungchoi95 commented on docs/Questions.md:
```diff
@@ -0,0 +1,11 @@
+## 미션 3-1 질문 
+
+### 시스템에 저장하는 username은 어떻게 정할지?
```
저는 이미 플랫폼별로 registrationId가 구분되어있어, username을 prefix에 붙여서 사용하지 않고 별도 컬럼을 둘 것 같아요.

조회 쿼리를 하게 되었을 때 플랫폼을 무시하고 조회하기보다는 플랫폼에 따라 username을 별도로 조회할일이 많을 것이고, 말씀해주신 식별자가 같은 경우에 대해서도 사실 unique를 플랫폼정보와 같이 걸면 문제가 없을 것이구요.

prefix로 합쳐도 결국 username을 파싱해서 빼서 써야해서 불필요한 과정들을 거치게된다는 문제도 있을 수 있겠네요.

jinyoungchoi95 commented on docs/Questions.md:
```diff
@@ -0,0 +1,11 @@
+
+### 시스템에 저장하는 username은 어떻게 정할지?
+
+### OAuth2를 이용한 로그인 시, 임의로 생성된 username으로 로그인이 가능하게 할 것인가?  
```
제가 질문을 정확하게 이해하지는 못하였는데요.

OAuth2 인증과정이 완료된 후 그 정보를 바탕으로 회원가입된 정보는 임시가 아닌 앞으로도 계속 사용해야하는 유저 데이터로써 사용되어요. 별도로 추가회원가입이라는 절차는 없습니다.

회원 정보에 대한 업데이트가 필요하다면 그건 회원가입과는 별개로 로그인 후 사용자에게 요청할 별도의 행위일 것 같아요.

haero77 replied:
@jinyoungchoi95 

https://github.com/next-step/spring-security-oauth2/pull/19#discussion_r1976437913 
진영님 위 코멘트 확인해주시면 감사하겠습니다..!

원 질문을 요약하면 다음과 같습니다.

1. OAuth2 로그인 및 웹 서버에서 username 생성
2. 1에서 생성된 username에 대한 로그인을 허용해야할까?

지금은 비밀번호를 난수로 생성하고 있기 때문에 로그인은 불가능하겠지만, 
이와는 별개로 임의로 생성한 username에 대한 로그인을 원천적으로 차단하는게 맞을지에 대한 고민입니다..! 

저는 유저가 회원가입시 입력한 username이 아니고, 즉 유저가 인지하지 못하기 때문에 기본적으로 막아야한다고는 생각하지만, 실무에서라면 (비밀번호를 알 수 없으므로) 어차피 로그인이 불가능하니 별도의 조치는 취하지 않을 것 같긴합니다 ㅎㅎ

관련해서 진영님 생각은 어떠신지 궁금하네요..!

jinyoungchoi95 replied:
OAuth2로 생성된 유저의 경우 form에서 수행했던 username/password 라는 개념이 따로 없습니다. 토큰으로써 이미 인증되었다고 확인이 되는 것이고 그 토큰의 식별값을 db에 저장하는거죠.

어떤 값으로 생성을 할수도 없이 저장자체가 불가능하도록 막혀있다가 맞을 것 같아요.

jinyoungchoi95 commented on docs/Questions.md:
```diff
@@ -0,0 +1,11 @@
+
+### OAuth2를 이용한 로그인 시, 임의로 생성된 username으로 로그인이 가능하게 할 것인가?  
+
+### authentication 패키지와 provider 패키지 양방향 참조.
```
우선 구현체마다 Client가 분리되어있지 않기 때문에 실제 spring security 구조를 참고하여 수정하면서 확인해보면 좋을 것 같아요.

haero77 replied:
감사합니다. 이 부분은 2단계 진행하면서 반영해보겠습니다..!

jinyoungchoi95 commented on src/main/resources/application-local.yml:
```diff
@@ -0,0 +1,17 @@
+security:
+  oauth2:
+    github:
+      client-id: ${OAUTH2_GITHUB_CLIENT_ID}
```
client-id는 식별값이기 때문에 숨겨둘 필요는 없지 않을까요?

haero77 replied:
처음에는 일종의 개인정보가 아닐까 생각해서 보안을 위해 숨겨두었는데요, 
생각해보니 OAuth2 인증 페이지로 리다이렉트시에 이미 Query String으로 노출되고 있네요 😅

---

[a7f5d13](https://github.com/next-step/spring-security-oauth2/pull/19/commits/a7f5d13bbbf2549436b3ff7ea19c8cb8d2e939ef) 커밋에 반영되었습니다 :)

jinyoungchoi95 commented on src/main/java/nextstep/security/oauth2/provider/SecurityOAuth2Properties.java:
```diff
@@ -0,0 +1,104 @@
+import org.springframework.util.Assert;
+
+@ConfigurationProperties(prefix = "security.oauth2")
+public class SecurityOAuth2Properties {
```
Properties가 우선 Github와 Google로 구현체명이 명시되어있는데요. 이렇게 되는 경우 플랫폼이 추가될 때마다 운영코드가 변경되어야한다는 문제가 있습니다.

어떤 플랫폼을 사용하는지에 대해서는 프로덕션 코드에서는 모두 추상화가 되어있고, "github" "google"과 같은 정보가 코드에서 주로 나타나지는 않아요. yml에서만 관리하는 형태인거죠.

그래서
```java
public class OAuth2ClientProperties {

    private final Map<String, Provider> provider = new HashMap<>();
    private final Map<String, Registration> registration = new HashMap<>();
```
와 같은 추상화된 형태로 받아올 수 있도록 수정이 필요해보여요 :)

jinyoungchoi95 commented on src/main/java/nextstep/security/oauth2/authentication/ClientRegistration.java:
```diff
@@ -0,0 +1,11 @@
+import nextstep.security.oauth2.provider.OAuth2Provider;
+
+public record ClientRegistration(
+        OAuth2Provider provider,
```
이어서 `ClientRegistration`도 사실 `GitHubClient`과 `GoogleClient`을 가지기보다는 `ClientRegistration`이라는 정보자체만을 가진다고 봐주시면 되어요. 어떤 플랫폼인지에 대해서 별개로 구현체를 가질 필요는 없는거죠.
(이미 registrationId가 그 역할을 하고 있으니까요.)

https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistration.java

jinyoungchoi95 commented on src/main/java/nextstep/security/oauth2/provider/SecurityOAuth2Properties.java:
패키지에 대한 이야기를 하면, `SecurityOAuth2Properties`는 OAuth2에서 조금 더 상위의 OAuth2 자체 속성이라는 개념이기 때문에 provider보다는 oauth2 하위에 있는 것이 맞을 것 같아요.

`provider`와 `registration`은 별개의 개념입니다. sample의 yml 값을 보시면 조금 더 와닿으실 것 같아요.

jinyoungchoi95 commented on src/main/java/nextstep/security/oauth2/authentication/GitHubLoginRedirectFilter.java:
```diff
@@ -0,0 +1,57 @@
+
+import java.io.IOException;
+
+public class GitHubLoginRedirectFilter extends GenericFilterBean {
```
마찬가지로 플랫폼별로 redirectFilter가 별개로 구현되어있는 것이 아니라 `OAuth2AuthorizationRequestRedirectFilter`라는 필터가 모든 OAuth2 요청을 처리합니다.

jinyoungchoi95 commented on src/main/java/nextstep/security/oauth2/authentication/GitHubLoginRedirectFilter.java:
```diff
@@ -0,0 +1,57 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(GitHubLoginRedirectFilter.class);
+
+    private final RequestMatcher matcher = new MvcRequestMatcher(HttpMethod.GET, "/oauth2/authorization/github");
```
path에서 github라는 정보가 registrationId가 되고, `/oauth2/authorization/{regisrationId}`에서 그 값을 꺼내오는 형태로 사용하게되는거죠. 추출된 id를 가지고 그에 맞는 `ClientRegistration`를 사용하여 OAuth2 과정을 거친다고 봐주시면 좋을 것 같아요.

OAuth2는 모든 플랫폼들이 동일한 스펙을 가지고 있기 때문에 각각의 구현체를 따로 둘 필요는 없습니다. 표준 프로토콜이니까요.

jinyoungchoi95 commented on src/main/java/nextstep/security/oauth2/authentication/OAuth2LoginAuthenticationFilter.java:
```diff
@@ -0,0 +1,132 @@
+        try {
+            Authentication authResult = attemptAuthentication(request, new OAuth2AuthorizationCode(authorizationCode));
+
+            SecurityContext context = SecurityContextHolder.createEmptyContext();
+            context.setAuthentication(authResult);
+            SecurityContextHolder.setContext(context);
+            securityContextRepository.saveContext(context, request, response);
```
작업은 finally에서 수행되어도 좋겠네요 :)

haero77 replied:
[33af47b](https://github.com/next-step/spring-security-oauth2/pull/19/commits/33af47bf0202214b687798013db75c7799816f72) 커밋에 해당 내용 반영해봤습니다..! 

OAuth2 로그인 이후 더 이상 SecurityContext를 쓸만한 곳이 보이지 않아 `SecurityContextHolder.clearContext()`를 이용하여 컨텍스트를 정리했는데 혹시 이 부분을 의도하신 건지 궁금하네요 ㅎㅎ

jinyoungchoi95 replied:
엇 아니요 clear를 추가한 것을 의도한 것은 아니었고 지금처럼 context 저장만을 의미한 것이었어요.

jinyoungchoi95 commented on src/main/java/nextstep/security/oauth2/authentication/OAuth2LoginAuthenticationFilter.java:
```diff
@@ -0,0 +1,132 @@
+        return this.authenticationManager.authenticate(authRequest);
+    }
+
+    private OAuth2AuthenticationToken generateUnAuthenticatedToken(
```
마찬가지로 github, google이라는 정보는 여기서 분기처리될 필요없이 registrationId를 가지고 Map에서 값을 꺼내와서 사용할 수 있는 형태면 좋을 것 같아요.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 선호님 😄

OAuth2 미션을 함께하게 된 최진영입니다.

미션 잘 진행해주셨네요 👍 몇가지 코멘트 남겨두었으니 확인부탁드려요 😄

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

haero77 committed:
refactor: OAuth2 클라이언트 환경변수를 관리하는 클래스를 OAuth2ClientProperties로 리네이밍

haero77 committed:
refactor: OAuth2 제공자가 추가되어도 프로덕션 코드는 변경 없도록 환경변수를 추상화하여 관리

haero77 committed:
refactor: OAuth2 Provider에 맞게 authorization URL로 리다이렉트하는 책임을 OAuth2AuthorizationRequestRedirectFilter에서 담당

haero77 committed:
docs: update Requirements

haero77 committed:
feat: OAuth2 인증 이후 SecurityContext 초기화 과정 추가

haero77 reviewed (commented):
@jinyoungchoi95 진영님 상세한 리뷰 감사드립니다..! 덕분에 제가 고민했던 부분이 리뷰 덕분에 많이 해결되고 있네요 😆

OAuth2 제공자가 추가되어도 프로덕션 코드는 변경되지 않는 것에 초점을 맞추어 피드백 진행해봤습니다..! 

1. OAuth2ClientProperties의 구조 변경. Provider, Registration 정보를 나눠서 관리. ([f363b67](https://github.com/next-step/spring-security-oauth2/pull/19/commits/f363b67fd74b2589152e388291c54fe81df7c545))
2. OAuth2AuthorizationRequestRedirectFilter를 이용하여 제공자에 관계없이 authorization uri 리다이렉트를 한 곳에서 관리. ([e6c73ca](https://github.com/next-step/spring-security-oauth2/pull/19/commits/e6c73caf7eeed18d280b737e72af00cd318b7f7c))

크게 위 2가지의 작업을 진행했으며, 아직 기존의 GoogleClient & GitHubClient는 통합하지 못했는데 이 부분은 2단계 진행하면서 도전해보겠습니다..! 잘 부탁드려요...!


피드백 반영 범위는 [여기](https://github.com/next-step/spring-security-oauth2/pull/19/files/3181b44e1983f200e207b18c5c04232b571ad5b3..a7f5d13bbbf2549436b3ff7ea19c8cb8d2e939ef) 참고해주시면 감사하겠습니다...!

haero77 committed:
feat: OAuth2 Client ID를 yml 파일에 노출

- 식별값이므로 노출 상관없으며, OAuth2 프로토콜(RFC 6749)에 의해 OAuth2 제공자 인증 페이지로 리다이렉트시에 URL의 쿼리 스트링으로 이미 노출되고 있음.

jinyoungchoi95 reviewed (approved):
안녕하세요 선호님 😄

남은 내용은 말씀해주신대로 다음 단계에서 실제 구현체를 확인해보시면서 반영 부탁드려요 😄

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
